### PR TITLE
Sliding into the PRs

### PR DESCRIPTION
**Pull request by bretsky**
Best programmer NA

### CODE REVIEW

File: code/controller/astar.cpp [hunk 1]
```diff
@@ -227,6 +227,7 @@ void nrg::search_path(
     node *node_start = world[start.x()][start.y()];
     node *node_dest = world[dest.x()][dest.y()];
     astar_search_path(node_start, node_dest, world, node_path);
+    node_path.push_front(node_start);
```
Comment (Mogball) [inline code/controller/astar.cpp]: leave this out for now; we don't need the ObjectProcedure wasting time trying to get to the starting position

File: code/controller/astar.cpp [hunk 2]
```diff
@@ -394,7 +395,7 @@ array2d<int> nrg::grid_kernelize(
     weak_ref<GridDisplay> grid,
     weak_ref<param_manager> pm
 ) {
-    constexpr int wall = GridDisplay::DEFAULT_WEIGHT;
+    int wall = GridDisplay::default_weight();
```
Comment (Mogball) [inline code/controller/astar.cpp]: make `GridDisplay::default_weight()` a `constexpr` function and add all the `constexpr` back

File: code/controller/astar.cpp [hunk 3]
```diff
@@ -428,10 +429,10 @@ void nrg::scale_path_pixels(
     weak_ref<GridDisplay> grid,
     std::vector<vector2i> &path
 ) {
-    constexpr int b = GridDisplay::GRID_SIZE;
-    constexpr int b2 = b / 2;
-    int gx = grid->x();
-    int gy = grid->y();
+    int b = GridDisplay::grid_size();
```
Comment (Mogball) [inline code/controller/astar.cpp]: ditto with the `constexpr`

File: code/controller/astar.cpp [hunk 4]
```diff
@@ -428,10 +429,10 @@ void nrg::scale_path_pixels(
...
+    int b2 = b / 2;
+    double gx = grid->x();
+    double gy = grid->y();
```
Comment (Mogball) [inline code/controller/astar.cpp]: why `double`?  `vector2i` is a pair of `int`s

File: code/controller/astar.cpp [hunk 5]
```diff
@@ -470,9 +471,9 @@ static void wallify(
     array2d<int> &walls,
     std::vector<rect2i> &rects
 ) {
-    constexpr int wall = GridDisplay::DEFAULT_WEIGHT;
-    constexpr int b = GridDisplay::GRID_SIZE;
-    constexpr int b2 = b / 2;
```
Comment (Mogball) [inline code/controller/astar.cpp]: same deal with `constexpr`

File: code/gui/griddisplay.cpp [hunk 6]
```diff
@@ -15,30 +15,46 @@
...
+    MAX_BUTTONS_X = 50,
+    MAX_BUTTONS_Y = 50
+};
+
 const QString buttonStyle =
     // Clear button style
     "background-color: rgba(0, 0, 0, 0%);"
-    "width: 8px;"
-    "height: 8px;";
+    "width: 0px;"
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: make these `QString`s just macros

File: code/gui/griddisplay.cpp [hunk 7]
```diff
@@ -52,18 +68,62 @@ static void swap_rect_coords(int &x1, int &y1, int &x2, int &y2) {
...
+int GridDisplay::default_weight() {
+    return DEFAULT_WEIGHT;
+}
+
+int GridDisplay::grid_size() {
+    return GRID_SIZE;
+}
+
+double GridDisplay::scene_width() {
+    return m_scene->width();
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: this function isn't called anywhere

File: code/gui/griddisplay.cpp [hunk 8]
```diff
@@ -52,18 +68,62 @@ static void swap_rect_coords(int &x1, int &y1, int &x2, int &y2) {
...
+}
+
+double GridDisplay::scene_height() {
+    return m_scene->height();
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: neither is this one?

File: code/gui/griddisplay.cpp [hunk 9]
```diff
@@ -52,18 +68,62 @@ static void swap_rect_coords(int &x1, int &y1, int &x2, int &y2) {
...
+}
+
+double GridDisplay::parent_width() {
+    return m_image_viewer->width();
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: nor this one

File: code/gui/griddisplay.cpp [hunk 10]
```diff
@@ -52,18 +68,62 @@ static void swap_rect_coords(int &x1, int &y1, int &x2, int &y2) {
...
+}
+
+double GridDisplay::parent_height() {
+    return m_image_viewer->height();
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: or this one

File: code/gui/griddisplay.cpp [hunk 11]
```diff
@@ -52,18 +68,62 @@ static void swap_rect_coords(int &x1, int &y1, int &x2, int &y2) {
...
+}
+
+double GridDisplay::x() {
+    return m_view->x();
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: why return `double`? `m_view->x()` returns `int` I'm pretty sure

File: code/gui/griddisplay.cpp [hunk 12]
```diff
@@ -52,18 +68,62 @@ static void swap_rect_coords(int &x1, int &y1, int &x2, int &y2) {
...
+}
+
+double GridDisplay::y() {
+    return m_view->y();
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: ditto

File: code/gui/griddisplay.cpp [hunk 13]
```diff
@@ -52,18 +68,62 @@ static void swap_rect_coords(int &x1, int &y1, int &x2, int &y2) {
...
     m_scene = std::make_unique<QGraphicsScene>(this);
-    m_scene->setSceneRect(QRect(0, 0, SCENE_WIDTH, SCENE_HEIGHT));   //TODO: set bounding rectangle to m_img.size()
+    m_scene->setSceneRect(QRect(0, 0, SCENE_WIDTH, SCENE_HEIGHT));
+    
+    //TODO: set bounding rectangle to m_img.size()
+
     m_view = std::make_unique<QGraphicsView>(m_scene.get(), image_viewer);
     m_view->setStyleSheet("background: transparent");
-    m_view->setMinimumSize(SCENE_WIDTH, SCENE_HEIGHT);
+    m_view->setMinimumSize(0, 0);
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: is this because `ImageViewer` might be resizing?

File: code/gui/griddisplay.cpp [hunk 14]
```diff
@@ -135,6 +195,13 @@ void GridDisplay::clear_selection() {
 }
 
 void GridDisplay::show_grid() {
+    m_row_count = (m_image_viewer->height() / 2) / GRID_SIZE;
+    m_column_count = (m_image_viewer->width() / 2) / GRID_SIZE;
+    m_scene->setSceneRect(QRect(0, 0, m_column_count * GRID_SIZE, m_row_count * GRID_SIZE));
+    m_view->resize(m_column_count * GRID_SIZE, m_row_count * GRID_SIZE);
+    m_view->move(m_image_viewer->width() / 2 - m_scene->width() / 2, m_image_viewer->height() / 2 - m_scene->height() / 2);
+    m_button = array2d<GridButton *>(m_column_count, m_row_count);
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: these arrays should be initialized once in the constructor only

File: code/gui/griddisplay.cpp [hunk 15]
```diff
@@ -135,6 +195,13 @@ void GridDisplay::clear_selection() {
 }
 
 void GridDisplay::show_grid() {
+    m_row_count = (m_image_viewer->height() / 2) / GRID_SIZE;
+    m_column_count = (m_image_viewer->width() / 2) / GRID_SIZE;
+    m_scene->setSceneRect(QRect(0, 0, m_column_count * GRID_SIZE, m_row_count * GRID_SIZE));
+    m_view->resize(m_column_count * GRID_SIZE, m_row_count * GRID_SIZE);
+    m_view->move(m_image_viewer->width() / 2 - m_scene->width() / 2, m_image_viewer->height() / 2 - m_scene->height() / 2);
```
Comment (Mogball) [inline code/gui/griddisplay.cpp]: this line is kind of hard to read, could you add more local variables

File: code/gui/griddisplay.h [hunk 16]
```diff
@@ -21,23 +21,24 @@
...
-        SCENE_WIDTH = 100,      // Default: 800
-        SCENE_HEIGHT = 100,     // Default: 400
-        MAX_BUTTONS_X = 50,
-        MAX_BUTTONS_Y = 50
-    };
+    static int default_weight();
+    static int grid_size();
+
+    double scene_width();
+    double scene_height();
```
Comment (Mogball) [inline code/gui/griddisplay.h]: same comments about the functions here

File: code/gui/serialbox.cpp [hunk 17]
```diff
@@ -58,14 +58,32 @@ SerialBox::SerialBox(
     setup_power_slider(*(ui->left_slider));
     setup_power_slider(*(ui->right_slider));
     setup_slider(*(ui->delay_slider), Delay::DELAY_DEFAULT, Delay::DELAY_MIN, Delay::DELAY_MAX, Delay::DELAY_INTERVAL);
+    m_edit_boxes[0] = ui->up_box;
```
Comment (Mogball) [inline code/gui/serialbox.cpp]: use brace initialization in the constructor initialization list

`SerialBox() : ..., m_edit_boxes{ui->up_box, ...}` etc.

File: code/gui/serialbox.cpp [hunk 18]
```diff
@@ -58,14 +58,32 @@ SerialBox::SerialBox(
...
+    m_edit_boxes[1] = ui->down_box;
+    m_edit_boxes[2] = ui->left_box;
+    m_edit_boxes[3] = ui->right_box;
+    m_sliders[0] = ui->up_slider;
+    m_sliders[1] = ui->down_slider;
+    m_sliders[2] = ui->left_slider;
+    m_sliders[3] = ui->right_slider;
+
+    QDialogButtonBox *button_box = ui->button_box;
```
Comment (Mogball) [inline code/gui/serialbox.cpp]: is this variable really necessary?

File: code/gui/serialbox.cpp [hunk 19]
```diff
@@ -76,6 +94,76 @@ SerialBox::~SerialBox() {
...
+void SerialBox::right_slider_changed(int value) {
+    slider_changed(value, 3);
+}
+
+void SerialBox::box_changed(int value, int dir) {
+    m_sliders[dir]->setValue(value);
+}
+
+void SerialBox::up_box_changed() {
+    QString value = ui->up_box->text();
```
Comment (Mogball) [inline code/gui/serialbox.cpp]: these `*_box_changed()` functions have some repeated code

File: code/gui/serialbox.cpp [hunk 20]
```diff
@@ -76,6 +94,76 @@ SerialBox::~SerialBox() {
     delete ui;
 }
 
+void SerialBox::slider_changed(int value, int dir) {
+    m_edit_boxes[dir]->setText(QString::number(value));
+    switch (dir) {
+        case 0:
+            m_solenoid->change_up_power(value);
```
Comment (Mogball) [inline code/gui/serialbox.cpp]: why don't you call a generic solenoid change power function instead?

File: code/gui/serialbox.h [hunk 21]
```diff
@@ -33,11 +34,31 @@ Q_OBJECT
 
     Q_SLOT void append_text(const std::string &text);
 
+    Q_SLOT void up_slider_changed(int value);
```
Comment (Mogball) [inline code/gui/serialbox.h]: get rid of the spaces between similar functions

File: code/gui/serialbox.h [hunk 22]
```diff
@@ -5,6 +5,7 @@
 
 #include <QDialog>
 #include <QSlider>
+#include <QLineEdit>
```
Comment (Mogball) [inline code/gui/serialbox.h]: if you don't need the type definition in the header use a forward declaration
```c++
class QSlider;
class QLineEdit;
```

and then just include the appropriate header in the source file only

File: code/gui/serialbox.h [hunk 23]
```diff
@@ -33,11 +34,31 @@ Q_OBJECT
...
+
+    void box_changed(int value, int dir);
+
 private:
     enum Power {
-        POWER_INTERVAL = 8,
+        POWER_INTERVAL = 1,
         POWER_MAX = 255,
-        POWER_MIN = 0
+        POWER_MIN = 240
```
Comment (Mogball) [inline code/gui/serialbox.h]: If these don't need to be seen in the header put them in the source file

File: code/controller/astar.cpp (hunk 2 above)
Comment (bretsky) [reply code/controller/astar.cpp]: Aight if I do this then I need to put the function back in the `griddisplay.h` file, but since we moved the enums, I can't access the constants from the header. Should I move the enum back?

File: code/gui/griddisplay.cpp (hunk 13 above)
Comment (bretsky) [reply code/gui/griddisplay.cpp]: Yeah if you set that then it can't resize below the minimum size I believe

File: code/gui/griddisplay.cpp (hunk 14 above)
Comment (bretsky) [reply code/gui/griddisplay.cpp]: I wasn't able to make that work since we don't know the size of `m_image_viewer` yet so `m_row_count` and `m_column_count` are still 0 in the constructor.

File: code/controller/astar.cpp (hunk 2 above)
Comment (Mogball) [reply code/controller/astar.cpp]: ehh you're right forget about it

File: code/gui/griddisplay.cpp (hunk 6 above)
Comment (bretsky) [reply code/gui/griddisplay.cpp]: If I do this some of them need to be converted back to `QString`s later

File: code/gui/griddisplay.cpp (hunk 7 above)
Comment (bretsky) [reply code/gui/griddisplay.cpp]: Yeah I thought I was gonna need this stuff but it turned out it wasn't needed.

File: code/gui/griddisplay.cpp (hunk 6 above)
Comment (Mogball) [reply code/gui/griddisplay.cpp]: yeah that's fine, these don't need to be allocated in static memory

Comment (Mogball) [reply code/gui/griddisplay.cpp]: you can do `QString style(STYLE_MACRO)` whenever you need it

File: code/gui/serialbox.cpp (hunk 17 above)
Comment (bretsky) [reply code/gui/serialbox.cpp]: I don't want to because it segfaults and I'm gonna cry because it makes no sense

Comment (bretsky) [reply code/gui/serialbox.cpp]: Yeah if I initialise them there then the array ends up being filled with null and/or garbage. I don't know why.

Comment (Mogball) [reply code/gui/serialbox.cpp]: oh true tm

Comment (bretsky) [reply code/gui/serialbox.cpp]: Goddamn it Jeff
Just baited me hard, I knew there was a reason I didn't do it like that the first time.

Comment (Mogball) [reply code/gui/serialbox.cpp]: I didn't realise the pointers weren't setup until setupUi lol I thought that just assigns the widgets

File: code/gui/serialbox.cpp (hunk 20 above)
Comment (bretsky) [reply code/gui/serialbox.cpp]: I do but I call it in `solenoid.cpp` instead. Should I move the switch case there?

File: code/gui/serialbox.h (hunk 23 above)
Comment (bretsky) [reply code/gui/serialbox.h]: Both enums?

File: code/camera/cameradisplay.cpp [hunk 24]
```diff
@@ -63,16 +63,14 @@ CameraDisplay::CameraDisplay(QWidget *parent) :
...
 
     // Setup rotation slider
-    m_ui->rotate_slider->setTickInterval(Rotation::ROTATION_INTERVAL);
-    m_ui->rotate_slider->setTickPosition(QSlider::TicksBelow);
-    m_ui->rotate_slider->setMaximum(Rotation::ROTATION_MAX);
-    m_ui->rotate_slider->setMinimum(Rotation::ROTATION_MIN);
+    setup_slider(*(m_ui->rotate_slider), 0, Rotation::ROTATION_MIN, Rotation::ROTATION_MAX, Rotation::ROTATION_INTERVAL, true);
+
+    // Setup griddisplay sliders
+    setup_slider(*(m_ui->horizontal_grid_slider), 0, GridLocation::GRIDLOCATION_MIN, GridLocation::GRIDLOCATION_MAX, GridLocation::GRIDLOCATION_INTERVAL, true);
```
Comment (Mogball) [inline code/camera/cameradisplay.cpp]: These lines are a bit long 
Can you keep them under 150 characters ish

File: code/camera/cameradisplay.cpp [hunk 25]
```diff
@@ -201,3 +201,30 @@ void CameraDisplay::pressed_play(bool checked) {
     Q_EMIT toggle_rotation(checked);
     m_ui->play_button->setText(checked ? "⏸" : "▶");
 }
+
+void CameraDisplay::setup_slider(QSlider &slider, int value, int min, int max, int interval, bool tracking) {
```
Comment (Mogball) [inline code/camera/cameradisplay.cpp]: This line too

File: code/camera/cameradisplay.h [hunk 26]
```diff
@@ -35,6 +36,8 @@ Q_OBJECT
      */
     int get_weighting();
 
+    void setup_slider(QSlider &slider, int value, int min, int max, int interval, bool tracking);
```
Comment (Mogball) [inline code/camera/cameradisplay.h]: And this xd

File: code/camera/cameradisplay.h [hunk 27]
```diff
@@ -235,6 +246,12 @@ Q_OBJECT
         ROTATION_MAX = 180
     };
 
+    enum GridLocation {
+        GRIDLOCATION_INTERVAL = 10,
```
Comment (Mogball) [inline code/camera/cameradisplay.h]: Do these need to be in the header?

File: code/gui/serialbox.h [hunk 28]
```diff
@@ -33,23 +36,27 @@ Q_OBJECT
...
+    void slider_changed(int value, int dir);
 
+    void box_changed(int dir);
+
+private:
     Ui::SerialBox *ui;
     std::shared_ptr<Solenoid> m_solenoid;
 
+    QSlider *m_sliders[4];
+    QLineEdit *m_edit_boxes[4];
```
Comment (Mogball) [inline code/gui/serialbox.h]: If these aren't dynamically allocated can you replace them with weak_ref? You can do this in a different PR

File: code/camera/cameradisplay.cpp (hunk 24 above)
Comment (bretsky) [reply code/camera/cameradisplay.cpp]: Just put in newlines? The actual code itself is fine right?

Comment (Mogball) [reply code/camera/cameradisplay.cpp]: yeah like
```c++
function_call(
    argument1, argument2, argument3,
    argument4, agument5, argument6
);
```

Comment (bretsky) [reply code/camera/cameradisplay.cpp]: This is why text editors have word wrap tho :cry:
